core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java

```diff
@@ -0,0 +1,140 @@
+     * @param onResult a function that takes the cluster state update acknowledged flag and timed out flag as parameters
+     *                 and returns a response to be sent on the listener
+     * @param onFailure executed on failure of the cluster state update, with the Throwable passed as input
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public <T> ActionListener<ClusterStateUpdateResponse> wrapUpdateListenerWithWaiting(final CreateIndexClusterStateUpdateRequest request,
+                                                                                        final ActionListener<T> actionListener,
+                                                                                        final BiFunction<Boolean, Boolean, T> onResult,
+                                                                                        final Consumer<Throwable> onFailure) {
+        return new ActionListener<ClusterStateUpdateResponse>() {
+            @Override
+            public void onResponse(ClusterStateUpdateResponse response) {
+                if (response.isAcknowledged()) {
+                    final String indexName = request.index();
+                    final ActiveShardCount waitForActiveShards = request.waitForActiveShards();
+                    final TimeValue timeout = request.masterNodeTimeout();
+                    try {
+                        // the cluster state update with the created index has been acknowledged, now wait for the
+                        // configured number of active shards to be allocated before returning, as that is when indexing
+                        // operations can take place on the newly created index
+                        if (waitForActiveShards == ActiveShardCount.NONE) {
+                            // not waiting, so just run whatever we were to run when the waiting is
+                            actionListener.onResponse(onResult.apply(true, false));
+                            return;
+                        }
+
+                        // wait for the configured number of active shards to be allocated before returning
+                        final ClusterStateObserver observer =
+                            new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+                        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+                            @Override
+                            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                                return waitForActiveShards.enoughShardsActive(newState, indexName);
```

Comment:
i will add a test to put “index.wait_for_active_shards” in the index settings with a value different than the default, then ensure it used that value found in the settings